### PR TITLE
Refactor route logic

### PR DESCRIPTION
**Pull request by cammace**
This PR is intended to fix a number of issues, mainly around the internal method monitorStepProgress which was difficult to write unit test for. It has now been broken up into separate methods that have one specific job, this is allowing me to write more unit test.

cc: @mapbox/navigation

### CODE REVIEW

**Comment by cammace** [comment]
Refactoring's mostly complete, just need to fix up checkstyle issues and document the classes/methods. 

@zugaldia ready for the review so we can do the `0.1.1` release.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/MapboxNavigation.java [hunk 1]
```diff
@@ -47,10 +49,10 @@
   private boolean isBound;
 
   // Navigation variables
-  private CopyOnWriteArrayList<AlertLevelChangeListener> alertLevelChangeListeners;
-  private CopyOnWriteArrayList<NavigationEventListener> navigationEventListeners;
-  private CopyOnWriteArrayList<ProgressChangeListener> progressChangeListeners;
-  private CopyOnWriteArrayList<OffRouteListener> offRouteListeners;
+  private List<AlertLevelChangeListener> alertLevelChangeListeners;
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/MapboxNavigation.java]: @cammace Why the change? Any drawbacks from keeping the previous type?

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java [hunk 2]
```diff
@@ -36,6 +33,7 @@
  * @since 0.1.0
  */
 @Experimental
+@SuppressWarnings( {"MissingPermission"})
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: Instead of attaching this to the whole class, let's put on the line(s) where it's actually need to avoid masking other instances.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java [hunk 3]
```diff
@@ -79,8 +77,7 @@ public void setupNotification(Activity activity) {
     // Sets up the top bar notification
     notifyBuilder = new NotificationCompat.Builder(this)
       .setContentTitle("Mapbox Navigation")
-      .setContentText("Distance: " + routeProgress.getCurrentLegProgress().getCurrentStepProgress()
-        .getDistanceRemaining())
+      .setContentText("navigating")
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: Can we provide a more descriptive text by default? Is there an API for the dev to update this text? (can be ticketed separately)

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/SnapLocation.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    // Pass in the snapped location with all the other location data remaining intact for their use.
+    location.setLatitude(routeProgress.getUsersCurrentSnappedPosition().getLatitude());
+    location.setLongitude(routeProgress.getUsersCurrentSnappedPosition().getLongitude());
+
+    location.setBearing(snapUserBearing());
+
+    return location;
+  }
+
+  // TODO split this method
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/SnapLocation.java]: Let's ticket or remove if done.

File: navigation/libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/RouteProgressTest.java [hunk 5]
```diff
@@ -14,66 +14,66 @@
...
-
-    assertEquals(nextRouteProgress.getCurrentLeg(), route.getLegs().get(0));
-    assertEquals(nextRouteProgress.getAlertUserLevel(), LOW_ALERT_LEVEL);
-    assertEquals(nextRouteProgress.getLegIndex(), 0);
-    assertEquals(nextRouteProgress.getDistanceRemaining(), 4037.6, 0.1);
-    assertEquals(nextRouteProgress.getDistanceTraveled(), 279.8, 0.1);
-    assertEquals(nextRouteProgress.getDurationRemaining(), 856, DELTA);
-    assertEquals(nextRouteProgress.getFractionTraveled(), 0.0648, 0.0001);
-  }
+//  private static final int NONE_ALERT_LEVEL = 0;
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/RouteProgressTest.java]: What's gonna happen with all these tests?

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/LocationUpdatedThread.java [hunk 6]
```diff
@@ -1,349 +0,0 @@
...
-    double userSnapToStepDistanceFromManeuver = RouteUtils.getDistanceToNextStep(
-      truePosition,
-      routeProgress.getCurrentLeg(),
-      routeProgress.getCurrentLegProgress().getStepIndex()
-    );
-
-    double secondsToEndOfStep = userSnapToStepDistanceFromManeuver / location.getSpeed();
-    boolean courseMatchesManeuverFinalHeading = false;
-
-    // TODO set to eventually adjust for different direction profiles.
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/LocationUpdatedThread.java]: Let's ticket or remove if done.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java [hunk 7]
```diff
@@ -125,61 +122,57 @@ public NavigationService getService() {
...
     }
-
-    responseHandler = new Handler();
-    locationUpdatedThread = new LocationUpdatedThread(responseHandler, options);
-    locationUpdatedThread.start();
-    locationUpdatedThread.getLooper();
-    Timber.d("Background thread started");
   }
 
+  // TODO use this to update notification
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: Let's ticket or remove if done.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java [hunk 8]
```diff
@@ -125,61 +122,57 @@ public NavigationService getService() {
...
         for (NavigationEventListener navigationEventListener : navigationEventListeners) {
           navigationEventListener.onRunning(true);
         }
       }
-
+      if (locationEngine.getLastLocation() != null) {
+        navigationEngine.onLocationChanged(directionsRoute, locationEngine.getLastLocation());
+      }
     } else {
+      // TODO throw exception here
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: Let's ticket or remove if done.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java [hunk 9]
```diff
@@ -193,14 +186,6 @@ public void endNavigation() {
...
-    responseHandler.removeCallbacksAndMessages(null);
-
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
-      locationUpdatedThread.quitSafely();
-    } else {
-      locationUpdatedThread.quit();
-    }
-
     // Lower accuracy to minimize battery usage while not in navigation mode.
     // TODO restore accuracy state to what user had before nav session
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: Let's ticket or remove if done.

File: navigation/libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/RouteControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+  @Test
+  public void isUserArriving_alertLevelShouldBeMedium() throws Exception {
+    RouteProgress routeProgress
+      = new RouteProgress(route, Mockito.mock(Location.class), 0, 10, NavigationConstants.MEDIUM_ALERT_LEVEL);
+    int alertLevel = RouteController.isUserArriving(NavigationConstants.MEDIUM_ALERT_LEVEL, routeProgress);
+    Assert.assertEquals(NavigationConstants.MEDIUM_ALERT_LEVEL, alertLevel);
+  }
+
+  // TODO improve increaseIndex for testing
```
Comment (zugaldia) [inline navigation/libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/RouteControllerTest.java]: Let's ticket or remove if done.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/MapboxNavigation.java (hunk 1 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/MapboxNavigation.java]: It seems unnecessary at least for now since we aren't passing these list to a different thread.

Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/MapboxNavigation.java]: @cammace We aren't but the developer might work with different threads?

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java (hunk 3 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: Planned on overhauling the notification stuff in a separate PR, The notification isn't even working right now during a navigation session...

Comment (zugaldia) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: @cammace Sounds good, let's ticket it separately.

Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: ticketed in https://github.com/mapbox/mapbox-navigation-android/issues/25

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java (hunk 7 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: noted on #25

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java (hunk 8 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: `NavigationException` has been added and is now being thrown here.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java (hunk 9 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/NavigationService.java]: ticketed in https://github.com/mapbox/mapbox-navigation-android/issues/38 and removed.

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/SnapLocation.java (hunk 4 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/SnapLocation.java]: ticketed in https://github.com/mapbox/mapbox-navigation-android/issues/39

File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/LocationUpdatedThread.java (hunk 6 above)
Comment (cammace) [inline navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/LocationUpdatedThread.java]: support for different direction profiles is already ticked in https://github.com/mapbox/mapbox-navigation-android/issues/24

**Comment by cammace** [comment]
@zugaldia ready for a second pass

**Comment by jom0000** [comment]
ผโนคอมเมน

**Comment by jom0000** [comment]
🇹 กลับมาประเทศกู 🇭
